### PR TITLE
add parallel odps table reader

### PR DESCRIPTION
**Pull request by QiJune**
This PR support both ODPS data reading and data preprocessing with multi-processes. You can use the ParallelTableReader in your custom reader.

```python
from elasticdl.python.data.parallel_table_reader import ParallelTableReader
class CustomParallelDataReader(ODPSDataReader):
    def __init__(self, **kwargs):
        ODPSDataReader.__init__(self, **kwargs)

    def read_records(self, task):
        start = task.start
        end = task.end
        table = self._get_odps_table_name(task.shard_name)
        table = table.split(".")[1]
        batch_size = (end - start) // 4
        columns = self._kwargs.get("columns")
        project = self._kwargs["project"]
        access_id = self._kwargs["access_id"]
        access_key = self._kwargs["access_key"]
        endpoint = self._kwargs.get("endpoint")
        partition = self._kwargs.get("partition", "")
        num_parallel_processes = 4
        transform_fn = py_parse_data # your data preprocess fn
        pd = ParallelTableReader(access_id, access_key, project, endpoint, table, partition,
                               columns, batch_size, num_parallel_processes, transform_fn)
        pd.reset((start, end - start))
        shard_count = pd.get_shards_count()
        for i in range(shard_count):
            records = pd.get_records()
            for record in records:
                yield record
        pd.stop()

    @property
    def records_output_types(self):
        return tf.string

def custom_data_reader(data_origin, records_per_task=None, **kwargs):
    return CustomParallelDataReader(
        project=os.environ[MaxComputeConfig.PROJECT_NAME],
        access_id=os.environ[MaxComputeConfig.ACCESS_ID],
        access_key=os.environ[MaxComputeConfig.ACCESS_KEY],
        table=data_origin,
        endpoint=os.environ.get(MaxComputeConfig.ENDPOINT),
        records_per_task=records_per_task,
        **kwargs,
    )
```

It will speed up both data reading and data preprocessing greatly.

### CODE REVIEW

**Review by terrytangyuan — APPROVED** [review]
Can we merge the two ODPS readers?

**Comment by QiJune** [comment]
@terrytangyuan Yes, I will refactor ODPS reader related code in next PR.

**Comment by terrytangyuan** [comment]
@QiJune This PR introduces a new class that's unnecessary though. Why refactoring later?

**Comment by QiJune** [comment]
@terrytangyuan The ParallelTableReader introduces a `transform_fn` function, which puts data preprocessing part into the multi-processes too. It's very common that we will insert some data preprocessing  logic after reading/loading data. I am afraid the interface of our data reader will also be changed.

We have an internal experiment which needs this `ParallelTableReader` feature.  We could merge this PR first and refactor data reader design very soon.

File: elasticdl/python/data/parallel_table_reader.py [hunk 1]
```diff
@@ -0,0 +1,111 @@
+from multiprocessing import Process, Queue
+
+from odps import ODPS
+
+
+class ParallelTableReader(object):
```
Comment (brightcoder01) [inline elasticdl/python/data/parallel_table_reader.py]: ParallelTableReader => ParallelOdpsTableReader?

File: elasticdl/python/data/parallel_table_reader.py [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+        self._shard_idx = 0
+        self._worker_idx = 0
+
+        for i in range(self._num_parallel_processes):
+            index_queue = Queue()
+            self._index_queues.append(index_queue)
+
+            p = Process(target=self._worker_loop, args=(i,))
+            p.daemon = True
+            p.start()
```
Comment (brightcoder01) [inline elasticdl/python/data/parallel_table_reader.py]: Do we miss the code snippet to call p.join()?

File: elasticdl/python/data/parallel_table_reader.py [hunk 3]
```diff
@@ -0,0 +1,111 @@
+from multiprocessing import Process, Queue
+
+from odps import ODPS
+
+
+class ParallelTableReader(object):
+    def __init__(
```
Comment (brightcoder01) [inline elasticdl/python/data/parallel_table_reader.py]: I suggest to add more descriptions about the design on the Parallel Reader

File: elasticdl/python/data/parallel_table_reader.py (hunk 1 above)
Comment (QiJune) [reply elasticdl/python/data/parallel_table_reader.py]: Good idea, Done

File: elasticdl/python/data/parallel_table_reader.py (hunk 2 above)
Comment (QiJune) [reply elasticdl/python/data/parallel_table_reader.py]: Done

File: elasticdl/python/data/parallel_table_reader.py (hunk 3 above)
Comment (QiJune) [reply elasticdl/python/data/parallel_table_reader.py]: Done

**Pull request merged**